### PR TITLE
Phoenix

### PR DESCRIPTION
**Pull request by nelsonic**
+ [x] Ports Hits from Node.js to Phoenix #66

### CODE REVIEW

**Comment by nelsonic** [comment]
@RobStallion this is a pretty massive PR considering the app is so Tiny ... 🙄 
Please review and comment / merge when you can (_no rush ... just before you leave PT..._) 🌻 
Thanks! 🌈

File: lib/hits.ex [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  - conn: Map the standard Plug.Conn info see: hexdocs.pm/plug/Plug.Conn.html
+
+  Returns String with user-agent
+  """
+  def get_user_agent_string(conn) do
+    # TODO: sanitise useragent string https://github.com/dwyl/fields/issues/19
+    # extract user-agent from conn.req_headers:
+    [{_, ua}] = Enum.filter(conn.req_headers, fn {k, _} ->
+      k == "user-agent" end)
+    # IO.inspect(ua, label: "ua")
```
[the remark this answers is not in the document]
Comment (RobStallion) [reply lib/hits.ex]: There are a few of these throughout the code so I am assuming that they were meant to be left in. Please feel free to ignore this

File: test/coverage_test.exs [hunk 2]
```diff
@@ -0,0 +1,14 @@
+defmodule CoverageTest do
+  use ExUnit.Case
+  # use HitsWeb.ConnCase
+
+  # Annoyingly, Phoenix creates a bunch of functions that are not executed
```
Comment (RobStallion) [inline test/coverage_test.exs]: 👍 👍 👍 👍 👍 👍 👍 👍 👍 👍 👍 👍 👍 👍 👍 👍 👍 👍 👍

Comment (RobStallion) [reply test/coverage_test.exs]: I saw that you set up the `coveralls.json` to skip some files. Do you know know if it is possible to skip this file or if it is possible to skip a function in a file? I have had a look at the coveralls docs and have never seen any way of doing this (unless you moved this function to a different file which was ignored, and then called the function somewhere but that seems SUPER hacky)

File: test/hits_web/controllers/hit_controller_test.exs [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  test "GET /org/dashboard", %{conn: conn} do
+    res = put_req_header(conn, "user-agent", "Hackintosh")
+     |> put_req_header("accept-language", "en-GB,en;q=0.5")
+     |> get("/org/dashboard")
+
+    # IO.inspect(res.resp_body, label: "res.body")
+    assert res.resp_body =~ ~s(Graphical dashboard coming soon!)
+  end
+
+  test "make_badge with default count 1" do
```
Comment (RobStallion) [inline test/hits_web/controllers/hit_controller_test.exs]: This test appears to be a duplicate of the one in `hits_test.exs`

File: CODE_OF_CONDUCT.md [hunk 4]
```diff
@@ -0,0 +1,46 @@
+# Contributor Covenant Code of Conduct
```
Comment (RobStallion) [inline CODE_OF_CONDUCT.md]: 👍 🎉

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,3 @@
+**Please read** our
```
Comment (RobStallion) [inline CONTRIBUTING.md]: 🥇

**Comment by RobStallion** [comment]
@nelsonic I left a few comments but nothing is even close to a blocker. Let me know if you want me to merge 👍

File: lib/hits.ex (hunk 1 above)
[the remark this answers is not in the document]
Comment (nelsonic) [reply lib/hits.ex]: Yeah, I've left the IO.inspect instances commented out and will remove them in a future iteration. 👍

File: lib/hits/hit.ex [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+  @doc false
+  def changeset(hit, attrs) do
+    hit
+    |> cast(attrs, [])
+    |> validate_required([])
+  end
+
+  def insert(attrs) do
+    # IO.inspect(attrs, label: "insert(attrs)")
```
[the remark this answers is not in the document]
Comment (nelsonic) [reply lib/hits/hit.ex]: Yeah, it's fine for now. Will remove in future.

File: lib/hits/hit.ex [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+    field :useragent_id, :id
+
+    timestamps()
+  end
+
+  @doc false
+  def changeset(hit, attrs) do
+    hit
+    |> cast(attrs, [])
+    |> validate_required([])
```
[the remark this answers is not in the document]
Comment (nelsonic) [reply lib/hits/hit.ex]: Yeah, will add items to the array on next iteration.

File: lib/hits/repository.ex [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+  @doc """
+  insert/1 inserts and returns the repository.id.
+
+  ## Parameters
+
+  - attrs: Map with the name of the person the repository belongs to.
+
+  returns Int user.id
+  """
+  def insert(attrs) do
```
[the remark this answers is not in the document]
Comment (nelsonic) [reply lib/hits/repository.ex]: @RobStallion I had all the functions in one file before. I don't mind either way. I only split them out to make the file smaller.

File: README.md [hunk 9]
```diff
@@ -1,259 +1,904 @@
...
 
-Install dependencies:
-```sh
-npm install
+Install elixir/node dependencies
+and setup Webpack static asset compilation (_with hot reloading_):
+
+```
+mix deps.get
+cd assets && npm install
```
[the remark this answers is not in the document]
Comment (nelsonic) [reply README.md]: Yeah, up a level is good to add to this command chain. 👍

File: test/coverage_test.exs (hunk 2 above)
Comment (nelsonic) [reply test/coverage_test.exs]: Nah, I don't think excoveralls has that level of granularity.

File: test/hits_web/controllers/hit_controller_test.exs (hunk 3 above)
Comment (nelsonic) [reply test/hits_web/controllers/hit_controller_test.exs]: Indeed. We can safely remove one of them.

**Comment by nelsonic** [comment]
@RobStallion if you feel it's worth tidying up the items you've noted, please go for it.
Otherwise, please merge.
Thanks! ☀️

**Comment by RobStallion** [comment]
@nelsonic I'll have a look at creating the context I mentioned. I'll assign you when it's ready 👍

**Comment by RobStallion** [comment]
Actually, on second thought that is in no way at all a blocker so I am going to merge this now and I'll create a separate issue for that which I'll work on 👍

**Pull request merged**